### ISSUE TITLE
Realm: NCCL point-to-point communication incompatible with CUDA hijack

### ISSUE THREAD

**Issue by manopapad** [StanfordLegion/legion#969]
NCCL v2 introduced a feature called "groups", whereby instead of blocking after each operation, multiple operations can be collected in a group and fired together. This feature is useful for avoiding deadlocks when one thread is managing multiple devices, and for making point-to-point communication easy to program (e.g. the user can implement an all-to-all by simply posting one send to each other gpu and one recv from each other gpu, without worrying about the ordering).

A group is implemented by spawning multiple pthreads that each wait on a different communication, and push the results to the GPUs appropriately. This model is incompatible with the CUDA hijack, since these pthreads won't run on GPU processors, and thus won't have access to a device.

Since every task running on a GPU processor only has access to one device, NCCL collectives can be used safely in Legion without grouping, as long as we launch exactly one task per GPU (e.g. see https://gitlab.com/StanfordLegion/legion/-/blob/master/examples/nccl/nccl_legion_demo.cu). This method, however, is not sufficient for point-to-point communication; for that we still need to launch each operation on a separate thread that will execute asynchronously. AFAIU this doesn't have to be a pthread; we could instead launch a separate task for each send and recv, and Realm would context-switch away the tasks that get blocked.

With that, I can see a couple of possible solutions:

1. Somehow intercept the pthread launches happening inside NCCL and make sure they run on a GPU processor, with access to the appropriate device.

2. Modify NCCL to spawn tasks instead of pthreads (perhaps NCCL could be parameterized on a threading library). 

3. When using NCCL point-to-point communication in Legion, explicitly spawn a separate task for each send/recv call. This would be the simplest alternative, however it turns out that send & recv [internally invoke groupStart/groupEnd](https://github.com/NVIDIA/nccl/blob/master/src/collectives/sendrecv.cc#L20-L22), and thus any use of send or recv will always spawn a pthread, even if it's not needed; this behavior within NCCL would need to change (if the NCCL devs agree it's safe to do so).
